### PR TITLE
Add support for friendly routes in pages

### LINKED ISSUE #6325: Add support for friendly routes in pages
From https://github.com/aspnet/Mvc/issues/5353

Associated with each page is an application model type similar to `ControllerModel`. We'd build on top of this to provide simple runtime configuration of routes. For instance, to configure an additional route to the `/Catalog/Books.cshtml`, here's what the Startup code might look like:

```
services
    .AddMvc()
    .AddRazorPages(options =>
    {
        options.MapPageRoute("Books/{id?}", "/Catalog/Books.cshtml");
    });
```

### PR DESCRIPTION
**Pull request by pranavkm**
FIxes #6325

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 1]
```diff
@@ -170,6 +170,50 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
         public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options, string folderPath) =>
             AuthorizeFolder(options, folderPath, policy: string.Empty);
 
+        /// <summary>
+        /// Associates a friendly rotue with the page at the specified <paramref name="path"/>.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: route

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 2]
```diff
@@ -170,6 +170,50 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
...
+        /// </summary>
+        /// <param name="options">The <see cref="RazorPagesOptions"/>.</param>
+        /// <param name="path">The page path.</param>
+        /// <param name="routeUrl">The route to associate with the page.</param>
+        /// <returns>The <see cref="RazorPagesOptions"/>.</returns>
+        public static RazorPagesOptions MapPageRoute(this RazorPagesOptions options, string path, string routeUrl)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: We should revisit this again in p3, we don't in general want to put extension methods directly on options

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: routeURL is in the right name, it's `route` or `routeTemplate`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 3]
```diff
@@ -170,6 +170,50 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
...
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(path));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: This should be a page name not a path https://github.com/aspnet/Mvc/blob/ccdaa5a7291683c4b8fd26f898848a40cfe0698e/test/WebSites/RazorPagesWebSite/StartupWithBasePath.cs#L19

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 4]
```diff
@@ -170,6 +170,50 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
...
+            }
+
+            if (string.IsNullOrEmpty(routeUrl))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(routeUrl));
+            }
+
+            options.Conventions.Add(new PageConvention(path, model =>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: Is this supposed to add or supress the page's innate route(s)? It's not clear from the naming

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 5]
```diff
@@ -170,6 +170,50 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
...
+            {
+                // Use the route specified in MapPageRoute for outbound routing.
+                foreach (var selector in model.Selectors)
+                {
+                    selector.AttributeRouteModel.SuppressLinkGeneration = true;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: This is probably the most surprising combination of what to do

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: `MapPageRoute` in System.Web was additive. We could call it `AddPageRoute` if we want to be super clear, but I wasn't sure if @DamianEdwards wanted to keep the method signature.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: Essentially what we do with `/Index` and `/` routes, no? Should we call this out in the summary \ remarks of this method?

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: This is not System.Web. We need to have a design that works for what we're trying to achieve, no one is going to be directly porting code from web forms -> pages

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs [hunk 6]
```diff
@@ -162,14 +162,62 @@ public static RazorPagesOptions AuthorizeFolder(this RazorPagesOptions options,
...
+        /// <para>
+        /// The page can be routed via <paramref name="route"/> in addition to the default set of path based routes.
+        /// All links generated for this page will use the specified route.
+        /// </para>
+        /// </summary>
+        /// <param name="options">The <see cref="RazorPagesOptions"/>.</param>
+        /// <param name="pageName">The page name.</param>
+        /// <param name="route">The route to associate with the page.</param>
+        /// <returns>The <see cref="RazorPagesOptions"/>.</returns>
+        public static RazorPagesOptions AddPageRoute(this RazorPagesOptions options, string pageName, string route)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/RazorPagesOptionsExtensions.cs]: Renamed to `AddPageRoute`

**Pull request closed without merging**